statistics/handle/lock_stats_handler.go

```diff
@@ -49,161 +29,50 @@ var (
-			}
-			tableLocked = append(tableLocked, tid)
-		} else {
-			skippedTables = append(skippedTables, tables[i].Schema.L+"."+tables[i].Name.L)
-		}
-	}
-
-	// Insert related partitions while don't warning duplicate partitions.
-	for _, tid := range pids {
-		if !isTableLocked(tableLocked, tid) {
-			if err := insertIntoStatsTableLocked(ctx, exec, tid); err != nil {
-				return "", err
-			}
-			tableLocked = append(tableLocked, tid)
-		}
-	}
-
-	// Update handle.tableLocked after transaction success, if txn failed, tableLocked won't be updated.
-	h.tableLocked = tableLocked
-
-	msg := generateSkippedTablesMessage(tids, skippedTables, lockAction, lockedStatus)
-	// Note: defer commit transaction, so we can't use `return nil` here.
-	return msg, err
-}
-
-func generateSkippedTablesMessage(tids []int64, dupTables []string, action, status string) string {
-	if len(dupTables) > 0 {
-		tables := strings.Join(dupTables, ", ")
-		var msg string
-		if len(tids) > 1 {
-			if len(tids) > len(dupTables) {
-				msg = fmt.Sprintf("skip %s %s tables: %s, other tables %s successfully", action, status, tables, status)
-			} else {
-				msg = fmt.Sprintf("skip %s %s tables: %s", action, status, tables)
-			}
-		} else {
-			msg = fmt.Sprintf("skip %s %s table: %s", action, status, tables)
-		}
-		return msg
-	}
-
-	return ""
-}
-
-func insertIntoStatsTableLocked(ctx context.Context, exec sqlexec.SQLExecutor, tid int64) error {
-	_, err := exec.ExecuteInternal(ctx, "INSERT INTO mysql.stats_table_locked (table_id) VALUES (%?) ON DUPLICATE KEY UPDATE table_id = %?", tid, tid)
-	if err != nil {
-		logutil.BgLogger().Error("error occurred when insert mysql.stats_table_locked", zap.String("category", "stats"), zap.Error(err))
-		return err
-	}
-	return nil
+	h.mu.RLock()
+	defer h.mu.RUnlock()
+	return lockstats.AddLockedTables(h.mu.ctx.(sqlexec.SQLExecutor), tids, pids, tables)
 }
 
-// LoadLockedTables load locked tables from store
-func (h *Handle) LoadLockedTables() error {
+// RemoveLockedTables remove tables from table locked array.
+// - tids: table ids of which will be unlocked.
+// - pids: partition ids of which will be unlocked.
+// - tables: table names of which will be unlocked.
+// Return the message of skipped tables and error.
+func (h *Handle) RemoveLockedTables(tids []int64, pids []int64, tables []*ast.TableName) (string, error) {
 	h.mu.Lock()
 	defer h.mu.Unlock()
-	ctx := kv.WithInternalSourceType(context.Background(), kv.InternalTxnStats)
-	exec := h.mu.ctx.(sqlexec.SQLExecutor)
-
-	tableLocked, err := loadLockedTables(ctx, exec, maxChunkSize)
-	if err != nil {
-		return err
-	}
 
-	// Update all locked tables.
-	h.tableLocked = tableLocked
-
-	return nil
+	return lockstats.RemoveLockedTables(h.mu.ctx.(sqlexec.SQLExecutor), tids, pids, tables)
 }
 
 // IsTableLocked check whether table is locked in handle with Handle.Mutex
-func (h *Handle) IsTableLocked(tableID int64) bool {
+func (h *Handle) IsTableLocked(tableID int64) (bool, error) {
```

Comment:
Can we allow this method to accept multiple table IDs like `IsTableLocked(tableIDs ...int64) ([]bool, error)` to get a better performance?

Reply:
Good point. I tried it this morning. It makes the code complicated. I will try it again.

Reply:
Addressed in [`a6a0bf3` (#46555)](https://github.com/pingcap/tidb/pull/46555/commits/a6a0bf31b0559451da544c5ace98819dce977a3f)